From: Iain Lane <laney@debian.org>
Date: Fri, 6 Nov 2020 13:33:13 +0000
Subject: Allow background-color and color on VteTerminal widgets to be
 customised via the theme

Yaru and Ambiance/Radiance need this. Add fallback style class and
move the CSS rules to it so that they can be overridden.

Forwarded: not-needed
Last-Update: 2021-11-08
---
 src/vtegtk.cc | 28 ++++++++++++++++++++++------
 1 file changed, 22 insertions(+), 6 deletions(-)

diff --git a/src/vtegtk.cc b/src/vtegtk.cc
index 521f745..a86be3d 100644
--- a/src/vtegtk.cc
+++ b/src/vtegtk.cc
@@ -92,6 +92,7 @@ template<typename T>
 constexpr bool check_enum_value(T value) noexcept;
 
 struct _VteTerminalClassPrivate {
+        GtkStyleProvider *fallback_style_provider;
         GtkStyleProvider *style_provider;
 };
 
@@ -893,6 +894,9 @@ try
 	_vte_debug_print(VTE_DEBUG_LIFECYCLE, "vte_terminal_init()\n");
 
         context = gtk_widget_get_style_context(&terminal->widget);
+        gtk_style_context_add_provider (context,
+                                        VTE_TERMINAL_GET_CLASS (terminal)->priv->fallback_style_provider,
+                                        GTK_STYLE_PROVIDER_PRIORITY_FALLBACK);
         gtk_style_context_add_provider (context,
                                         VTE_TERMINAL_GET_CLASS (terminal)->priv->style_provider,
                                         VTE_TERMINAL_CSS_PRIORITY);
@@ -2609,6 +2613,24 @@ vte_terminal_class_init(VteTerminalClass *klass)
 
         klass->priv = G_TYPE_CLASS_GET_PRIVATE (klass, VTE_TYPE_TERMINAL, VteTerminalClassPrivate);
 
+        klass->priv->fallback_style_provider = GTK_STYLE_PROVIDER (gtk_css_provider_new ());
+        /* Some themes don't define text_view_bg */
+        gtk_css_provider_load_from_data (GTK_CSS_PROVIDER (klass->priv->fallback_style_provider),
+                                         "@define-color text_view_bg @theme_base_color;\n"
+                                         "VteTerminal, " VTE_TERMINAL_CSS_NAME " {\n"
+#if (VTE_GTK == 4) || ((VTE_GTK == 3) && GTK_CHECK_VERSION (3, 24, 22))
+                                         "background-color: @text_view_bg;\n"
+#else
+                                         "background-color: @theme_base_color;\n"
+#endif
+                                         "color: @theme_text_color;\n"
+                                         "}\n",
+                                         -1
+#if VTE_GTK == 3
+                                         , NULL
+#endif
+                                         );
+
         klass->priv->style_provider = GTK_STYLE_PROVIDER (gtk_css_provider_new ());
 #if VTE_GTK == 3
         auto err = vte::glib::Error{};
@@ -2619,12 +2641,6 @@ vte_terminal_class_init(VteTerminalClass *klass)
         gtk_css_provider_load_from_data (GTK_CSS_PROVIDER (klass->priv->style_provider),
                                          "VteTerminal, " VTE_TERMINAL_CSS_NAME " {\n"
                                          "padding: 1px 1px 1px 1px;\n"
-#if (VTE_GTK == 4) || ((VTE_GTK == 3) && GTK_CHECK_VERSION (3, 24, 22))
-                                         "background-color: @text_view_bg;\n"
-#else
-                                         "background-color: @theme_base_color;\n"
-#endif
-                                         "color: @theme_text_color;\n"
                                          "}\n",
                                          -1
 #if VTE_GTK == 3
